src/App.bonus.js

```diff
@@ -0,0 +1,56 @@
+import React from "react";
+import { toast, ToastContainer } from "react-toastify";
+
+import { Todo } from "./Todo";
+
+import "../node_modules/react-toastify/dist/ReactToastify.css";
+import "./App.css";
+
+class App extends React.Component {
+  state = {
+    input: "",
+    todos: [],
+  };
+
+  onSubmit = (event) => {
+    event.preventDefault();
+    this.setState({
+      todos: [...this.state.todos, this.state.input],
+    });
+  };
+
+  removeTodo = (index) => {
+    this.setState({
+      todos: [...this.state.todos.slice(0, index), ...this.state.todos.slice(index + 1)],
```

Comment:
Just curious why `slice` instead of `filter`?

Reply:
Human words: `.filter` goes through the entire array even after it finds the right one. Example: if the one you're looking for is the first and the list has 9001 items, it'll still loop through the remaining 9000.

Math words: `.filter` is guaranteed O(N) even though the optimal case is O(N / 2) on average. It's not an order of magnitude different but is still worse.

Reply:
So `slice` does not need to actually go through all of the items? We know the indices we care about so it can just rearrange the references or something like that?

Reply:
Oh I see what you're saying: actually yeah I guess you're right 😄 . I'm used to thinking of these in terms of theoretical immutable arrays, where slice wouldn't need to go through - but yeah now that you mention it in practical terms these are roughly equivalent.

I don't know, it's always felt weird to me to use a `.filter` when you just want to check the one index...